pyo3-derive-backend/src/from_pyobject.rs

```diff
@@ -0,0 +1,531 @@
+                FieldAttribute::GetAttr(None) => quote!(getattr(stringify!(#ident))),
+                FieldAttribute::GetItem(Some(key)) => quote!(get_item(#key)),
+                FieldAttribute::GetItem(None) => quote!(get_item(stringify!(#ident))),
+            };
+            fields.push(quote!(#ident: obj.#ext_fn?.extract()?));
+        }
+        quote!(Ok(#self_ty{#fields}))
+    }
+
+    fn check_transparent_len(fields: &Fields) -> Result<()> {
+        if fields.len() != 1 {
+            return Err(spanned_err(
+                fields,
+                "Transparent structs and variants can only have 1 field",
+            ));
+        }
+        Ok(())
+    }
+}
+
+/// Attributes for deriving FromPyObject scoped on containers.
+#[derive(Clone, Debug, PartialEq)]
+enum ContainerAttribute {
+    /// Treat the Container as a Wrapper, directly extract its fields from the input object.
+    Transparent,
+    /// Change the name of an enum variant in the generated error message.
+    ErrorAnnotation(String),
+}
+
+impl ContainerAttribute {
+    /// Return whether this attribute is `Transparent`
+    fn transparent(&self) -> bool {
+        match self {
+            ContainerAttribute::Transparent => true,
+            _ => false,
+        }
+    }
+
+    /// Convenience method to access `ErrorAnnotation`.
+    fn annotation(&self) -> Option<&String> {
+        match self {
+            ContainerAttribute::ErrorAnnotation(s) => Some(s),
+            _ => None,
+        }
+    }
+
+    /// Parse valid container arguments
+    ///
+    /// Fails if any are invalid.
+    fn parse_attrs(value: &[Attribute]) -> Result<Vec<Self>> {
+        let mut attrs = Vec::new();
+        let list = get_pyo3_meta_list(value)?;
+        for meta in list.nested {
+            if let syn::NestedMeta::Meta(metaitem) = &meta {
+                match metaitem {
+                    Meta::Path(p) if p.is_ident("transparent") => {
+                        attrs.push(ContainerAttribute::Transparent)
+                    }
+                    Meta::NameValue(nv) if nv.path.is_ident("annotation") => {
+                        if let syn::Lit::Str(s) = &nv.lit {
+                            attrs.push(ContainerAttribute::ErrorAnnotation(s.value()))
+                        } else {
+                            return Err(spanned_err(&nv.lit, "Expected string literal."));
+                        }
+                    }
+                    other => {
+                        return Err(spanned_err(
+                            other,
+                            "Expected `transparent` or `annotation = \"name\"`",
+                        ))
+                    }
+                }
+            } else {
+                return Err(spanned_err(
+                    meta,
+                    "Unknown container attribute, expected `transparent` or \
+                    `annotation(\"err_name\")`",
+                ));
+            }
```

Comment:
How about dropping the else outside and doing the `Err` return just once? 

Also, I'd be tempted to not list the possible options here as it can get easily outdated.

```suggestion
                    _ => {} // return Err below
                }
            }
            
            return Err(spanned_err(
                meta,
                "Unrecongised `pyo3` container attribute",
            ));
```